trace-utils/src/send_data.rs

```diff
@@ -0,0 +1,229 @@
+    Any(anyhow::Error),
+}
+
+pub struct SendDataResult {
+    pub last_result: anyhow::Result<Response<Body>>,
+    pub requests_count: u64,
+    pub responses_count_per_code: HashMap<u16, u64>,
+    pub errors_timeout: u64,
+    pub errors_network: u64,
+    pub errors_status_code: u64,
+}
+
+impl SendDataResult {
+    fn new() -> SendDataResult {
+        SendDataResult {
+            last_result: Err(anyhow!("No requests sent")),
+            requests_count: 0,
+            responses_count_per_code: Default::default(),
+            errors_timeout: 0,
+            errors_network: 0,
+            errors_status_code: 0,
+        }
+    }
+
+    async fn update(
+        &mut self,
+        res: Result<Response<Body>, SendRequestError>,
+        expected_status: StatusCode,
+    ) {
+        self.requests_count += 1;
+        match res {
+            Ok(response) => {
+                *self
+                    .responses_count_per_code
+                    .entry(response.status().as_u16())
+                    .or_default() += 1;
+                self.last_result = if response.status() == expected_status {
+                    Ok(response)
+                } else {
+                    self.errors_status_code += 1;
+
+                    let body_bytes = hyper::body::to_bytes(response.into_body()).await;
+                    let response_body = String::from_utf8(body_bytes.unwrap_or_default().to_vec())
+                        .unwrap_or_default();
+                    Err(anyhow::format_err!(
+                        "Server did not accept traces: {response_body}"
+                    ))
+                }
+            }
+            Err(e) => match e {
+                SendRequestError::Hyper(e) => {
+                    if e.is_timeout() {
+                        self.errors_timeout += 1;
+                    } else {
+                        self.errors_network += 1;
+                    }
+                    self.last_result = Err(anyhow!(e));
+                }
+                SendRequestError::Any(e) => {
+                    self.last_result = Err(e);
+                }
+            },
+        }
+    }
+
+    fn error(mut self, err: anyhow::Error) -> SendDataResult {
+        self.last_result = Err(err);
+        self
+    }
+}
+
+fn construct_agent_payload(tracer_payloads: Vec<pb::TracerPayload>) -> pb::AgentPayload {
+    pb::AgentPayload {
+        host_name: "".to_string(),
+        env: "".to_string(),
+        agent_version: "".to_string(),
+        error_tps: 60.0,
+        target_tps: 60.0,
+        tags: HashMap::new(),
+        tracer_payloads,
+        rare_sampler_enabled: false,
+    }
+}
+
+fn serialize_proto_payload<T>(payload: &T) -> anyhow::Result<Vec<u8>>
+where
+    T: prost::Message,
+{
+    let mut buf = Vec::with_capacity(payload.encoded_len());
+    payload.encode(&mut buf)?;
+    Ok(buf)
+}
+
+#[derive(Debug, Clone)]
+pub struct SendData {
+    pub tracer_payloads: Vec<pb::TracerPayload>,
```

Comment:
Do we really need to expose `tracer_payloads` publicly? From what I can tell, it's only for the tests in the `trace-mini-agent`, which is not a good pattern. We shouldn't be exposing internal implementation details just for unit tests in this way. 

If we are only exposing `tracer_payloads` for tests, could we refactor the tests to use httpmock and properly test the outputs? If that is something that we want to explore, I think we could do that in a follow-up PR since we want to get this merged in soon. Maybe we can just add a TODO comment then?

Reply:
This is a good point. We shouldn't expose internals unnecessarily in this way, but I think that going all the way to sending to the `httpmock` isn't necessary. Just adding a test only method on the `SendData` implementation that hands out payload number X, is fine by me.

Reply:
I struggled with this because [coalesce_send_data](https://github.com/DataDog/libdatadog/blob/luc/sidecar-collect-trace_api-metrics/sidecar/src/service/tracing/trace_flusher.rs#L232) needs to mutate the `tracer_payloads` fields so I thought about implementing a getter to return a mutable reference but in the end is kind of the same as making the the field public. Also I thought about restructuring the SendData and TraceFlusher to improve the encapsulation but I realized it'd overcomplicate this PR. Anyway if you guys think the getter is a better solution we can go that way or if you can think of a better solution please let me know.

Reply:
That can be public to the crate, right?